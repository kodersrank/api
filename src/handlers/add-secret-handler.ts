import type { Handler } from 'elysia';
import { db } from '../db';
import type { Secret } from '../types/secret.interface';

export const addSecretHandler: Handler = async ({ body }) => {
  const { secret, expireAfterViews, expireAfter } = body as Secret;

  // Insert secret into the database
  const [id] = await db('secrets')
    .insert({
      secret,
      expireAfterViews,
      expireAfter,
      createdAt: new Date(),
      expiresAt:
        expireAfter > 0 ? new Date(Date.now() + expireAfter * 60000) : null,
    })
    .returning('id');

  const response = {
    hash: id,
    secretText: secret,
    createdAt: new Date().toISOString(),
    expiresAt:
      expireAfter > 0
        ? new Date(Date.now() + expireAfter * 60000).toISOString()
        : null,
    remainingViews: expireAfterViews,
  };

  return response;
};
