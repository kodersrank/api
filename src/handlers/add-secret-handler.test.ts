import {
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
  type Mock,
} from 'vitest';
import { addSecretHandler } from './add-secret-handler';
import type { Handler } from 'elysia';
import type { Knex } from 'knex';

vi.mock('../db', () => {
  const insert = vi.fn();

  return {
    insertMock: insert,
    db: vi.fn(() => ({
      insert,
    })),
  };
});
describe('addSecretHandler', () => {
  let db: Mock<Knex<any, unknown[]>>;

  let insertMock: Mock;

  const DEF_MOCK = {
    secret: 'ðŸ§‚',
    curDateRef: 1234567890123,
    expireAfterViews: 5,
    expireAfter: 3,
  };

  const mockData = {
    ...structuredClone(DEF_MOCK),
    get expiryDateRef(): number {
      return this.curDateRef + this.expireAfter * 60000;
    },
  };

  beforeEach(async () => {
    ({ db, insertMock } = (await import('../db')) as any);

    vi.spyOn(Date, 'now').mockImplementation(() => mockData.curDateRef);
  });

  afterEach(() => {
    vi.clearAllMocks();

    Object.assign(mockData, DEF_MOCK);
  });

  describe('When called', () => {
    let set: Record<string, unknown>;

    const SHA_256_HASH = expect.stringMatching(/^[a-f\d]{64}$/);

    let latestResponse: any;

    beforeEach(async () => {
      set = { status: null };

      latestResponse = await addSecretHandler({
        body: {
          secret: mockData.secret,
          expireAfterViews: mockData.expireAfterViews,
          expireAfter: mockData.expireAfter,
        },
        set,
      } as Parameters<Handler>[0]);
    });

    describe('insert statement', () => {
      it('should point to secrets table', () => {
        expect(db.mock.calls).toEqual([['secrets']]);
      });

      it('should be called', () => {
        expect(insertMock).toHaveBeenCalledOnce();
      });

      describe('when expireAfter value is positive', () => {
        it('should be provided a matching payload with valid expiry date', () => {
          const expectedCall = [
            expect.objectContaining({
              secret: mockData.secret,
              expire_after: mockData.expireAfter,
              expire_after_views: mockData.expireAfterViews,
              created_at: new Date(mockData.curDateRef),
              expires_at: new Date(mockData.expiryDateRef),
              hash: SHA_256_HASH,
            }),
          ];

          expect(insertMock.mock.calls).toEqual([expectedCall]);
        });

        describe('when expireAfter value is zero', () => {
          beforeAll(() => {
            mockData.expireAfter = 0;
          });

          it('should be provided a matching payload with valid expiry date', () => {
            const expectedCall = [
              expect.objectContaining({
                secret: mockData.secret,
                expire_after: mockData.expireAfter,
                expire_after_views: mockData.expireAfterViews,
                created_at: new Date(mockData.curDateRef),
                expires_at: null,
                hash: SHA_256_HASH,
              }),
            ];

            expect(insertMock.mock.calls).toEqual([expectedCall]);
          });
        });
      });
    });

    describe('HTTP status code', () => {
      it('should be set to 201', () => {
        expect(set.status).toBe(201);
      });
    });

    describe('when expireAfter is positive', () => {
      describe('Response body', () => {
        it('should match expected contract with expiry date', () => {
          const expected = expect.objectContaining({
            hash: SHA_256_HASH,
            secretText: mockData.secret,
            createdAt: new Date(mockData.curDateRef),
            expiresAt: new Date(mockData.expiryDateRef),
            expireAfter: mockData.expireAfter,
            expireAfterViews: mockData.expireAfterViews,
            remainingViews: mockData.expireAfterViews,
          });

          expect(latestResponse).toEqual(expected);
        });

        it('should match snapshot', () => {
          expect(latestResponse).toMatchSnapshot();
        });
      });
    });

    describe('when expireAfter is zero', () => {
      beforeAll(() => {
        mockData.expireAfter = 0;
      });

      describe('Response body', () => {
        it('should match expected contract without expiry date', () => {
          const expected = expect.objectContaining({
            hash: SHA_256_HASH,
            secretText: mockData.secret,
            createdAt: new Date(mockData.curDateRef),
            expiresAt: null,
            expireAfter: 0,
            expireAfterViews: mockData.expireAfterViews,
            remainingViews: mockData.expireAfterViews,
          });

          expect(latestResponse).toEqual(expected);
        });

        it('should match snapshot', () => {
          expect(latestResponse).toMatchSnapshot();
        });
      });
    });
  });
});
