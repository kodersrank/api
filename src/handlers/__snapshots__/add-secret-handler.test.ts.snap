// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addSecretHandler > When called > when expireAfter is positive > Response body > should match snapshot 1`] = `
{
  "createdAt": 2009-02-13T23:31:30.123Z,
  "expireAfter": 3,
  "expireAfterViews": 5,
  "expiresAt": 2009-02-13T23:34:30.123Z,
  "hash": "2a133810adc5429b038e170ee75b0c1b51eb0bc48e66e52fc6199231175dabd6",
  "remainingViews": 5,
  "secretText": "ðŸ§‚",
}
`;

exports[`addSecretHandler > When called > when expireAfter is zero > Response body > should match snapshot 1`] = `
{
  "createdAt": 2009-02-13T23:31:30.123Z,
  "expireAfter": 3,
  "expireAfterViews": 5,
  "expiresAt": 2009-02-13T23:34:30.123Z,
  "hash": "2a133810adc5429b038e170ee75b0c1b51eb0bc48e66e52fc6199231175dabd6",
  "remainingViews": 5,
  "secretText": "ðŸ§‚",
}
`;
